torch/csrc/api/include/torch/detail/any.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <torch/csrc/utils/memory.h>
+#include <torch/csrc/utils/variadic.h>
+
+#include <ATen/Error.h>
+
+#include <memory>
+#include <typeinfo>
+#include <utility>
+
+namespace torch {
+class Any;
+namespace detail {
+template <typename T>
+Any make_any(T&& value);
+} // namespace detail
+} // namespace torch
+
+namespace torch {
+/// A simplified implementation of `std::any` or `boost::any` which stores a
+/// type erased object, whose concrete value can be retrieved at runtime by
+/// checking if the `typeid()` of a requested type matches the `typeid()` of the
+/// object stored. It is simplified in that it does not handle copying, as we do
+/// not require it for our use cases. Moves are sufficient.
```

Comment:
I'm not sure how I feel about the fact that we're making our own implementation of `std::any`. A lot of details to get right when doing this.

Reply:
I looked quite closely at implementations of `any` in boost and libc++: https://github.com/llvm-mirror/libcxx/blob/master/include/any

I think the benefit of rolling our own is that we can lose a lot of the complexity of the implementation because we don't need many of the features, such as the ability to copy/clone any values. I also didn't see any nice, standard conformant open source implementation of `std::any`

Reply:
Also, if we want to provide a no RTTI path one day, we'll have more freedom to add it to our own implementation. But no doubt that rolling our own is always sketchy

Reply:
Can we copy over the boost::any implementation? Your argument holds true as long as the Any type is only used here, but I'll probably have a use case in c10 too, which might need a slightly different feature set. Having a more feature-complete implementation closer to the standard might do us some good in the future. Also because its behavior is more well-known and predictable than hand-rolling our own.